actors/builtin/miner/miner_actor.go

```diff
@@ -1097,315 +1470,331 @@ func (a Actor) OnDeferredCronEvent(rt Runtime, payload *CronEventPayload) *adt.E
-				randomnessEpoch := minEpoch(deadline.PeriodEnd(), rt.CurrEpoch()-ElectionLookback)
-				assignmentSeed := rt.GetRandomness(crypto.DomainSeparationTag_WindowedPoStDeadlineAssignment, randomnessEpoch, nil)
-				err = AssignNewSectors(deadlines, info.WindowPoStPartitionSectors, newSectors, assignmentSeed)
-				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to assign new sectors to deadlines")
+	epochReward := requestCurrentEpochBlockReward(rt)
+	pwrTotal := requestCurrentTotalPower(rt)
+
+	hadEarlyTerminations := false
+
+	powerDelta := PowerPair{big.Zero(), big.Zero()}
+	newlyVested := big.Zero()
+	penaltyTotal := abi.NewTokenAmount(0)
+	pledgeDelta := abi.NewTokenAmount(0)
 
-				// Store updated deadline state.
-				err = st.SaveDeadlines(store, deadlines)
-				builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to store new deadlines")
+	var st State
+	rt.State().Transaction(&st, func() interface{} {
+		var err error
+		{
+			// Vest locked funds.
+			// This happens first so that any subsequent penalties are taken from locked pledge, rather than free funds.
+			newlyVested, err = st.UnlockVestedFunds(store, rt.CurrEpoch())
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to vest funds")
+		}
+
+		// Record whether or not we _had_ early terminations in the queue before this method.
+		// That way, don't re-schedule a cron callback if one is already scheduled.
+		hadEarlyTerminations = havePendingEarlyTerminations(rt, &st)
+
+		// Note: because the cron actor is not invoked on epochs with empty tipsets, the current epoch is not necessarily
+		// exactly the final epoch of the deadline; it may be slightly later (i.e. in the subsequent deadline/period).
+		// Further, this method is invoked once *before* the first proving period starts, after the actor is first
+		// constructed; this is detected by !dlInfo.PeriodStarted().
+		// Use dlInfo.PeriodEnd() rather than rt.CurrEpoch unless certain of the desired semantics.
+		dlInfo := st.DeadlineInfo(currEpoch)
+		if !dlInfo.PeriodStarted() {
+			return nil // Skip checking faults on the first, incomplete period.
+		}
+		deadlines, err := st.LoadDeadlines(store)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadlines")
+		deadline, err := deadlines.LoadDeadline(store, dlInfo.Index)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadline %d", dlInfo.Index)
+		partitions, err := deadline.PartitionsArray(store)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load partitions for deadline %d", dlInfo.Index)
+		quant := st.QuantEndOfDeadline()
+
+		{
+			// Detect and penalize missing proofs.
+			faultExpiration := dlInfo.Close + FaultMaxAge
+			penalizePowerTotal := big.Zero()
+
+			for i := uint64(0); i < partitions.Length(); i++ {
+				key := PartitionKey{dlInfo.Index, i}
+				proven, err := deadline.PostSubmissions.IsSet(i)
```

Comment:
This is potentially slow. We should probably just expand this into a map (it's not _too_ large). We can also do this with bitfield magic, but that's probably more complicated than it's worth.